File: src/evai/monitor/EventMonitor.java
package evai.monitor;

import arc.*;
import arc.util.*;
import mindustry.game.*;
import java.util.*;

public class EventMonitor {
    private static final Set<String> activeListeners = new HashSet<>();
    private static final Map<String, Class<?>> eventClasses = new HashMap<>();

    static {
        for (Class<?> clazz : EventType.class.getClasses()) {
            String name = clazz.getSimpleName();
            if (name.endsWith("Event")) {
                eventClasses.put(name.substring(0, name.length() - "Event".length()), clazz);
            }
        }
    }

    public static boolean listen(String name) {
        Class<?> clazz = eventClasses.get(name);
        if (clazz == null) return false;
        if (activeListeners.contains(name)) return true;
        Events.on(clazz, e -> Log.info("[Eve] Event fired: " + name + "Event"));
        activeListeners.add(name);
        return true;
    }

    public static int listenAll() {
        int count = 0;
        for (String name : eventClasses.keySet()) {
            if (!activeListeners.contains(name)) {
                listen(name);
                count++;
            }
        }
        return count;
    }

    public static void stop(String name) {
        activeListeners.remove(name);
    }

    public static void stopAll() {
        activeListeners.clear();
    }

    public static List<String> list() {
        List<String> result = new ArrayList<>(activeListeners);
        Collections.sort(result);
        return result;
    }
}

File: src/evai/monitor/EventMonitor.kt
package evai.monitor

import arc.*
import arc.util.*
import mindustry.game.*
import java.util.concurrent.*

object EventMonitor {
    private val listeners = mutableMapOf<String, () -> Unit>()
    private val eventClasses = mutableMapOf<String, Class<*>>()

    init {
        for (clazz in EventType::class.java.classes) {
            val name = clazz.simpleName.removeSuffix("Event")
            eventClasses[name] = clazz
        }
    }

    fun listen(name: String): Boolean {
        val eventClass = eventClasses[name]
            ?: run { Log.warn("[Eve] Unknown event: $name"); return false }
        if (listeners.containsKey(name)) return true // already listening
        val handler = {
            Log.info("[Eve] Event fired: ${name}Event")
        }
        Events.on(eventClass, handler)
        listeners[name] = handler
        return true
    }

    fun listenAll(): Int {
        var count = 0
        for ((name, clazz) in eventClasses) {
            if (!listeners.containsKey(name)) {
                val handler = { Log.info("[Eve] Event fired: ${name}Event") }
                Events.on(clazz, handler)
                listeners[name] = handler
                count++
            }
        }
        return count
    }

    fun stop(name: String) {
        listeners.remove(name)
    }

    fun stopAll() {
        listeners.clear()
    }

    fun list(): List<String> = listeners.keys.sorted()
}